infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogDataStore.java

```diff
@@ -1,119 +1,276 @@
+     * Any snapshot read before this may result in in-complete history.
+     */
+    private final AtomicReference<Long> globalCompactionMark;
+
+    /**
+     * Cached committed log tail, up to which the log is consolidated.
      */
-    private final AtomicLong startingAddress = new AtomicLong(Address.NON_ADDRESS);
+    private final AtomicReference<Long> committedTail;
+
     /**
-     * Cached tail segment
+     * Cached flag to indicate if required state transfer (state is lost)
+     * possibly due to reset operation.
      */
-    private final AtomicLong tailSegment = new AtomicLong(Address.NON_ADDRESS);
+    private final AtomicReference<Boolean> requireStateTransfer;
+
+    public StreamLogDataStore(KvDataStore dataStore) {
+        this.dataStore = dataStore;
+        this.headSegment =
+                new AtomicReference<>(dataStore.get(HEAD_SEGMENT_RECORD, Address.MAX));
+        this.tailSegment =
+                new AtomicReference<>(dataStore.get(TAIL_SEGMENT_RECORD, ZERO_ADDRESS));
+        this.globalCompactionMark =
+                new AtomicReference<>(dataStore.get(COMPACTION_MARK_RECORD, Address.NON_ADDRESS));
+        this.committedTail =
+                new AtomicReference<>(dataStore.get(COMMITTED_TAIL_RECORD, Address.NON_ADDRESS));
+        this.requireStateTransfer =
+                new AtomicReference<>(dataStore.get(REQUIRE_STATE_TRANSFER_RECORD, false));
+    }
 
     /**
-     * Return current cached tail segment or get the segment from the data store if not initialized
+     * Return the current tail segment.
      *
      * @return tail segment
      */
-    public long getTailSegment() {
-        if (tailSegment.get() == Address.NON_ADDRESS) {
-            tailSegment.set(dataStore.get(TAIL_SEGMENT_RECORD, ZERO_ADDRESS));
-        }
-
+    long getTailSegment() {
         return tailSegment.get();
     }
 
     /**
-     * Update current tail segment in the data store
+     * Update current tail segment in the data store.
      *
-     * @param newTailSegment updated tail segment
+     * @param newTailSegment new tail segment to update
      */
-    public void updateTailSegment(long newTailSegment) {
-        if (tailSegment.get() >= newTailSegment) {
-            log.trace("New tail segment less than or equals to the old one: {}. Ignore", newTailSegment);
-            return;
+    void updateTailSegment(long newTailSegment) {
+        if (update(TAIL_SEGMENT_RECORD, tailSegment, newTailSegment, (curr, next) -> next > curr)) {
+            log.debug("Updated tail segment to: {}", newTailSegment);
+        } else {
+            log.trace("New tail segment {} not greater than current, ignore.", newTailSegment);
         }
+    }
 
-        log.debug("Update tail segment to: {}", newTailSegment);
-        dataStore.put(TAIL_SEGMENT_RECORD, newTailSegment);
-        tailSegment.set(newTailSegment);
+    /**
+     * Reset the current tail segment.
+     */
+    void resetTailSegment() {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use explicit scoping instead of the default package private level](https://app.codacy.com/app/annymartinez.m/CorfuDB/pullRequest?prid=4651450)